### ISSUE TITLE
multiple users reporting duplicate images while signed in

### ISSUE THREAD

**Issue by vrooje** [zooniverse/Panoptes#1069]
Our 2 moderators on GZ Bars have both reported seeing duplicate images. I'm going to check the subject list for duplicates and will update but just wanted to flag this in case it's related to zooniverse/Panoptes#787 or something else that isn't duplicate subjects.

**Comment by vrooje** [comment]
Update: I've just checked the metadata in the subject download from the project and I can't find any duplicates in the image name, so I don't think this is a duplicate-subject issue.

**Comment by camallen** [comment]
Any talk links that indicate they have seen duplicate images?

**Comment by vrooje** [comment]
I don't think so, yet... the mods posted on Talk but didn't link to specific images. Will ask them to do so in the future.

**Comment by camallen** [comment]
@vrooje closing for now, please re-open if you have more info.

**Comment by vrooje** [comment]
https://www.zooniverse.org/#/projects/vrooje/galaxy-zoo-bar-lengths/talk/21/280?page=1&comment=1515

From @Capella05:
 > Just got this image to classify for the second time - 465197

**Comment by vrooje** [comment]
Also found the [subject comment thread](https://www.zooniverse.org/#/projects/vrooje/galaxy-zoo-bar-lengths/talk/32/565), which I suspect is redundant but am including it just in case.

**Comment by vrooje** [comment]
I've now got more information about this:

Strangely I *don't* have a duplicate in the classification export for Capella05 on subject id 465197, despite the fact that my export was requested more than 24 hours after the reported duplicate.

I do have a different duplicate for that user, however:

`username   subject_id   metadata`
`Capella05  {"464985":  {"started_at":"2015-07-01T19:23:02.756Z","user...`
`Capella05  {"464985":  {"started_at":"2015-07-07T19:16:23.117Z","user...`

And I have a total of 159 duplicates from various users. That's about 1% of the classifications. Examples (with dummy usernames):

`thisuser  {"458602":  {"started_at":"2015-06-01T16:57:55.437Z","user...`
`thisuser  {"458591":  {"started_at":"2015-06-01T16:58:00.380Z","user...`
`thisuser  {"458584":  {"started_at":"2015-06-01T16:58:05.496Z","user...`
`thisuser  {"458550":  {"started_at":"2015-06-01T16:58:10.840Z","user...`
[11 more classifications of non-duplicate subjects]
`thisuser  {"458602":  {"started_at":"2015-06-01T16:59:23.785Z","user...`
[2 non-duplicate classifications]
`thisuser  {"458591":  {"started_at":"2015-06-01T16:59:40.418Z","user...`
`thisuser  {"458584":  {"started_at":"2015-06-01T16:59:44.218Z","user...`
[2 non-duplicate classifications]
`thisuser  {"458550":  {"started_at":"2015-06-01T16:59:55.702Z","user...`

thisuser has done 30 classifications total. And, more recently, from a user who has done 18 classifications total:

`     user_name  subject_id                                           metadata`
[1 non-dup classification]
`thatuser  {"464953":  {"started_at":"2015-07-02T20:39:11.600Z","user...`
`thatuser  {"464954":  {"started_at":"2015-07-02T20:39:40.693Z","user...`
`thatuser  {"464955":  {"started_at":"2015-07-02T20:40:10.855Z","user...`
`thatuser  {"464956":  {"started_at":"2015-07-02T20:40:30.784Z","user...`
`thatuser  {"464957":  {"started_at":"2015-07-02T20:40:38.595Z","user...`
`thatuser  {"464958":  {"started_at":"2015-07-02T20:40:42.172Z","user...`
`thatuser  {"464959":  {"started_at":"2015-07-02T20:40:45.651Z","user...`
[3 non-dup classifications]
`thatuser  {"464953":  {"started_at":"2015-07-02T20:41:05.295Z","user...`
`thatuser  {"464954":  {"started_at":"2015-07-02T20:41:27.165Z","user...`
`thatuser  {"464955":  {"started_at":"2015-07-02T20:41:33.603Z","user...`
`thatuser  {"464956":  {"started_at":"2015-07-02T20:41:47.336Z","user...`
`thatuser  {"464957":  {"started_at":"2015-07-02T20:41:51.548Z","user...`
`thatuser  {"464958":  {"started_at":"2015-07-02T20:41:56.526Z","user...`
`thatuser  {"464959":  {"started_at":"2015-07-02T20:42:00.374Z","user...`

I also have duplicates from users who have done many more classifications overall, though not proportionally more duplicates. And not everyone who has done hundreds of classifications has done duplicate classifications.

**Comment by vrooje** [comment]
More details:

 - Some of the duplicates are from not-logged-in users, but most of them are from logged-in users.
 - The duplicates sometimes span days, and I have also seen them ~60s apart and once 14s apart (re: #1128)
 - Some duplicates are actually triplicates. In that case they are often, but not always, minutes or seconds apart. (I saw one that was 1 [1 hour] 2 [2 minutes] 3).
 - Many of the duplicates (about a third) have the same subject across multiple users, i.e. logged-in users A and B *both* classified subject 99 twice. In those cases the users' first viewing of the same subject isn't always on the same day, e.g. user B classified subject 99 for the first time 25 hours after user A classified subject 99 for the first time.
 - There is an example where 2 logged-in users on different OSs and classifying about 3 days apart are repeating the same subjects. One did 18 classifications ("thatuser" above) and the other did 14. They each repeated the same 6 subjects.

I've isolated the duplicates from the full export and grouped them together and I'm happy to send that over if the full data would be helpful.

**Comment by mschwamb** [comment]
Non-logged in, is a won't fix issue - see the conversation here #1127 - that's how the system was designed 

The 60s thing on my  ticket #1128 isn't important, there are classifications made in between. 

Merging from the info in my ticket since 

Yeah got a very recent one subject 487143 dweilant 2015-07-12 23:43:49 and 2015-07-12 23:45:07 - this is after is live for correctly accounting for the subjects - now subject has seen almost 50% of the live subjects if that matters but that still leaves about 4000 subjects to select from

user who's on the other spectrum of classifiers from the above in this post Audriusa for subject 484145 2015-07-12 07:24:10 and 2015-07-12 07:25:32

notice same thing for nathalieg69 on subject 484066 2015-07-01 03:55:37 and 2015-07-01 04:05:37

Worth noting this is still occurring after the fix for updating subject classification counts/retirement

**Comment by camallen** [comment]
@vrooje @mschwamb just letting you know we are working on this.

**Comment by mschwamb** [comment]
Thanks @camallen

**Comment by camallen** [comment]
Linked to #1176

Update, we've had no luck tracing this to any direct implemented behaviour (in code) but it could be the result of message ordering between clients and busy / fast API end points where subjects are not dequeued before the next request for subjects comes in. 

We will be modifying the dequeue behaviour to happen directly after you've been served subjects and not wait till you submit a classification. This should make any race condition harder to come by as the timing between these "race" messages will be much greater. 

I'm going to leave this open. If any more duplicate reports come in (especially after #1176 is fixed) then please make us aware on this issue.

**Comment by mschwamb** [comment]
Thanks for the update, and thanks for working on tracking it down. @camallen If you can tell me a time stamp when this fix should be online, I'll check it over the next few days after that and report back on P4: Terrains

**Comment by vrooje** [comment]
Thanks - will keep you posted if this keeps happening. If all we can do is minimize it instead of fix it we should make sure @ggdhines knows to explicitly check for duplicates in the data aggregation phase.

**Comment by camallen** [comment]
@vrooje pretty sure he already is as we always serve subjects to users in panoptes (we mark them seen / retired, etc) but allow them to keep classifying.

**Comment by mschwamb** [comment]
@camallen though that status is not marked in the raw csv dump - if duplicates and retired images seen again can be marked that would be something handy to include in the csv for those doing their data reduction

**Comment by camallen** [comment]
@mschwamb can you open a sperate issue for this? Perhaps reopen #1086 after reading https://github.com/zooniverse/Panoptes-Front-End/issues/368

**Comment by mschwamb** [comment]
@camallen I can't reopen. I'm not a collaborator on this repo. If the powers that be can add me I can start do that or if you want to reopen it I'll comment

**Comment by camallen** [comment]
Just comment and @ mention me

**Comment by camallen** [comment]
@mschwamb @vrooje code has been deployed about 18:00 today BST. Let's see how it goes for the next few days.

**Comment by mschwamb** [comment]
@camallen Still happening I think - I'm just looking at duplicates from 2015-07-30 00:00:00'  on P4T

username Uganalandia view subject 491799 at 2015-07-30 12:11:50 and 2015-07-30 12:14:46 

gaga7 saw subject 491810 at 2015-07-30 15:41:52 and 2015-07-30 16:08:10

4thplanet4444 saw subject 486319 at 2015-07-31 07:39:39 and 2015-07-31 08:25:08

**Comment by mschwamb** [comment]
only checked for 4thplanet4444 but there classification in between the repeat image

**Comment by edpaget** [comment]
I wonder if them happening close together means anything?

**Comment by camallen** [comment]
proxy that doesn't respect headers on the get subjects call?

**Comment by mschwamb** [comment]
also it repeats at a short frequency for non-logged in users as well  not sure if that's helpful or not.

**Comment by mschwamb** [comment]
Did some digging

Here's the list of duplicates  (where the duplicate happened on a logged-in user)I have in my database dump from classifications post 7/30/2015 00:00:00 

duplicate 485036
duplicate 485259
duplicate 485312
duplicate 485361
duplicate 485375
duplicate 485415
duplicate 485467
duplicate 485532
duplicate 485599
duplicate 485616
duplicate 485650
duplicate 485675
duplicate 485697
duplicate 485738
duplicate 485753
duplicate 485772
duplicate 485815
duplicate 485836
duplicate 485837
duplicate 485903
duplicate 485957
duplicate 485996
duplicate 486009
duplicate 486064
duplicate 486076
duplicate 486088
duplicate 486091
duplicate 486113
duplicate 486125
duplicate 486141
duplicate 486154
duplicate 486160
duplicate 486174
duplicate 486319
duplicate 486570
duplicate 486642
duplicate 487587
duplicate 491251
duplicate 491252
duplicate 491307
duplicate 491330
duplicate 491375
duplicate 491392
duplicate 491402
duplicate 491439
duplicate 491493
duplicate 491496
duplicate 491605
duplicate 491677
duplicate 491679
duplicate 491693
duplicate 491727
duplicate 491728
duplicate 491763
duplicate 491767
duplicate 491769
duplicate 491772
duplicate 491790
duplicate 491799
duplicate 491806
duplicate 491810
duplicate 491812
duplicate 491832
duplicate 491847
duplicate 491863
duplicate 491883
duplicate 491888
duplicate 491893
duplicate 491911
duplicate 491915
duplicate 491920
duplicate 491932
duplicate 491941
duplicate 491967
duplicate 492051
duplicate 492068
duplicate 492073

77 subjects out of 4473 classifications that came from logged-in volunteers so 2% of the classifications per day

**Comment by vrooje** [comment]
Okay, starting on 28th July (the day after the deploy), I count 492 duplicate classifications, out of just under 3,300 total classifications since that day, so that's about a 15% rate in that time frame.

The vast majority of them (424) were from one user, who was logged in. Of the other 11 users who had duplicates within this time, 10 were logged in. The next-highest number of duplicates is 20 from another logged-in user.

**Comment by vrooje** [comment]
More generally, since the duplicates began the fraction of a user's classifications that involve duplicate subjects varies, sometimes by quite a lot:

![screen shot 2015-08-04 at 14 41 59](https://cloud.githubusercontent.com/assets/2357126/9062023/7e48b906-3ab7-11e5-9e19-9193bcde4260.jpg)

**Comment by bruggsy** [comment]
Hi all, experiencing the same problem. I have a dataset of 200, and trying to have some experts classify each subject for a "gold standard" dataset. Here are the splits I'm seeing:

| # Classifications | # Repeats Seen | Repeat % |
|-------------------|----------------|----------|
|               234 |            158 |     32.4 |
|               228 |            156 |     31.6 |
|                55 |             51 |      7.2 |

Obviously a small sample, hope it helps. For reference, all the users are logged in, I told them to refresh the page any time they saw an "already seen" ribbon. Also checked the subject report for duplicates, of which there are none. Anxious to see a fix, let me know if any more info is needed.

**Comment by vrooje** [comment]
I just got "already seen" on a project and refreshed and I got a new subject. Is that supposed to happen? For some reason I thought refreshing wasn't supposed to show a new subject.

**Comment by edpaget** [comment]
@camallen I think there's another possible area of inconsistency here: https://github.com/zooniverse/Panoptes/blob/facc095fbf1fa51066313384fc5c44e1fae1bf6e/lib/subject_selector.rb#L82. Were we can setup the Enqueue worker, dequeue a set of 10 subjects for the User to classify, but between the User classifying the subjects (and the subjects getting added to the UserSeenSubject list), the Enqueue Worker runs and adds some of the 10 subjects served to the user back onto their queue. 

Since we no longer dequeue a subject after a classification (since that happens in the original subject request), we can end up queueing duplicate subjects.

**Comment by edpaget** [comment]
@vrooje We've changed the behaviour to remove subjects from the queue after they are served to a user, but before they get classified to try to eliminate one of the causes of duplicate subjects.

**Comment by camallen** [comment]
@edpaget - good one. should we just dequeue on both events?

**Comment by edpaget** [comment]
Yes, but and we should make sure the on classification dequeue is done in the request cycle. 

Also I wonder if there's some condition we're missing that could be causing `workflow.finished?` or `user.has_finished?` to be falsy returning true and triggering project wide-selection?

**Comment by camallen** [comment]
I don't think so i think the queues are not de-populating properly, adding some error messaging here #1251 and i'm going to empty all queues today to get some baseline behaviour start point.

**Comment by mschwamb** [comment]
@camallen  and @edpaget do you want us go and check our classifications after these changes are live over the next few days and report back duplicates?

**Comment by camallen** [comment]
@mschwamb you can but most of the recent work was around getting more error reporting into our logs for debugging. There are still some dups coming into the queues that #1257 should help with. Once we iron these remaining bugs out we can look again. 

That said you should start to see greater spread of subjects coming out of the queue so #1163 should be resolved.

**Comment by camallen** [comment]
Ok - finally an update. I think we've got a handle on the issue now, with better reporting and a check to remove any duplicates before they get enqueued. There is still more work to come on a better underlying implementation in the future. 

Duplicates for logged in users should not be occurring from late last night / early today (14th Aug GMT) onwards. Feel free to look at data from this point to see what is happening and please let us know if you're still experiencing the issue.

**Comment by vrooje** [comment]
Just had a look at the classifications over the weekend in GZ Bar Lengths.

Total duplicate count: 0.
:)

**Comment by camallen** [comment]
@vrooje Happy days! thanks for the report. @bruggsy  and @mschwamb any chance you can have a look at the data to see if you are experiencing duplicates?

**Comment by mschwamb** [comment]
I checked late last night my time - none from logged in users and one for non-logged in (I think this fixes also might have helped with that rate as well).

**Comment by camallen** [comment]
Great news - I think we can do better for non-logged in by tracking subject ids on the front end.

**Comment by mschwamb** [comment]
@camallen that would be great.

**Comment by camallen** [comment]
I'm going to close this issue as I think we've got this sorted. In summary our queuing code had some bugs and some legacy use case was causing the queues to grow very very large, which allowed dups in. 

If anyone finds duplicate classifications happening since the latest date quoted above please comment here / re-open the issue asap.

**Comment by chrislintott** [comment]
+1

**Comment by vrooje** [comment]
Sadly, I have to re-open this. There were a handful of duplicates in GZ:BL between the closing of this issue and the sending of a newsletter to recruit people to GZ:BL, but since the recruitment there have been about 1,150, out of 36,000, so a duplicate percentage of about 3%.

800 of the duplicates were from not-logged-in users. I looked into this because another top user reported duplicates, but (as with Capella05's report previously) I couldn't find any recent duplicates in the database from this user.

Happy to provide more info as needed...

**Comment by camallen** [comment]
@vrooje 800 from non-logged in is by design in the api, i've got an issue open in the front end to be smarter about ignoring these, https://github.com/zooniverse/Panoptes-Front-End/issues/1427.

So the final 350 i'll need some extra information. Are these all from power users in the long tail? Since commit https://github.com/zooniverse/Panoptes/commit/86938540822bd129807cdf424fba749712aa010f we may get to a situation where we have nothing in the queue and just select something to show the user, they should have seen a banner saying they've seen this before though....seems there are still some bugs to iron out on this one.

**Comment by CKrawczyk** [comment]
There were two more uses on talk that have reported duplicate images:

https://www.zooniverse.org/talk/18/115/?comment=19991
https://www.zooniverse.org/talk/18/115/?comment=22586

**Comment by bruggsy** [comment]
Also been noticing this problem when annotating any subject set with more than 10 members, while logged in.

**Comment by camallen** [comment]
@bruggsy, do you have reports / talk subject id's, screenshots? Can you please confirm that these are real duplicates and not the expected behaviour, that is the api will return a set of subjects when you have classified all of them. The client should mark the images as retired / seen before in the browser, e.g.

![seen_before](https://cloud.githubusercontent.com/assets/295329/8649722/ab9a56a8-2964-11e5-98d3-7e9c287f5ce2.png)

**Comment by vrooje** [comment]
Still getting occasional [comments](https://www.zooniverse.org/projects/vrooje/galaxy-zoo-bar-lengths/talk/22/582/?comment=28913) from users about duplicates. I think this is due to a failure to submit the classification the first time rather than a duplicate registered classification, but this still makes me uncomfortable because we don't really know how often this is happening, right?

**Comment by DarrenMcRoy** [comment]
Oh, so this isn't just a WildCam issue? https://github.com/zooniverse/wildcam-gorongosa/issues/192

**Comment by bruggsy** [comment]
@camallen Sorry for not responding, been busy with other projects. Not sure what you mean exactly, when I do an API request for a subject set / single subject I don't see those fields. The subjects I was having problems with were 1037498 - 1037697, and subject sets 2422 - 2440. The exact problem was getting "already seen" subjects before I had seen the whole set, as far as I can tell I wasn't getting the same image multiple times with no banner the latter times.

**Comment by camallen** [comment]
@vrooje i'm looking into the dup reports for GZ Bars. 

@bruggsy the system should be showing you unseen un retired images first, then falls back to unseen retired, then once you have completed them all it'll just pick some at random to show to you. Can you please check again and provide me some workflow / subject set id's to reproduce for your account?

**Comment by bruggsy** [comment]
@camallen Sure, the workflow is 861 and the subject set associated with it is 2592, project is 292.

**Comment by srallen** [comment]
I setup a project today for Intro2Astro using a very small subject set with 22 subjects and @JulieAnnKU reported this issue. She was served a subject with the already seen flag before having seen all 22. Project id `1616`, subject set id `2754`, workflow id `1064`

**Comment by camallen** [comment]
I've got another lead on real duplicates for users. Seems some of the background workers haven't been running and creating the req'd tracking details for what they've seen. Linked to #1517, ensuring that these workers run as soon as possible after failure decreases the window for duplicates to emerge.

**Comment by mschwamb** [comment]
Happening on Comet Hunters too - https://www.zooniverse.org/projects/mschwamb/comet-hunters/talk/84/25164?comment=51262

**Comment by aliburchard** [comment]
@camallen this seems to still be popping up on Wildcam Gorongosa - had a whole bunch reported mid/late December, and at least one person reported this week: https://www.zooniverse.org/projects/zooniverse/wildcam-gorongosa/talk/79/7125?page=3

**Comment by camallen** [comment]
thanks, there was a bug showing the proper already seen / retired tags from teh API that got fixed in SGL.  Also some of the issues in SGL would be leading to duplicates showing under heavy api loads. 

We've got new code out to fix these and some more to get out that will hopefully alleviate this.

**Comment by aliburchard** [comment]
@camallen just wanted to let you know this is still happening: https://www.zooniverse.org/projects/zooniverse/wildcam-gorongosa/talk/79/7125?comment=74235&page=4

**Comment by camallen** [comment]
Hmm - seems the messaging about already seen / retired was removed at the end of last month. See this, https://github.com/zooniverse/Panoptes-Front-End/issues/2170. I can only hope we aren't messaging correctly. 

I'm checking this user details to see what's happening.

**Comment by mschwamb** [comment]
This seems to still be happening - example on Planet Four Terrains user_id number 1328372 classified subject 1328372 twice within a ten minute span. This was on February 9th.

**Comment by mschwamb** [comment]
Volunteer is seeing 'already seen' flag on and off,  but there's ~6000 new images on Comet Hunters 
See the volunteers description [here](https://www.zooniverse.org/projects/mschwamb/comet-hunters/talk/210/40383?comment=94235)

**Comment by camallen** [comment]
closing this in favour of #1640 since we "are" labelling the duplicates as already seen, please put all reports onto the "Already Seens" issue from now ow.
